proof-of-spacetime.md

```diff
@@ -0,0 +1,198 @@
+
+```go
+// Verify a PoSt.
+func VerifyPoSt(sectorSize BytesAmount, sectors SectorSet, seed Seed, proof PoStProof, faults FaultSet) bool {
+    challenges := DerivePoStChallenges(seed, faults, sectorSize, len(sectors))
+    sectorsSorted := []
+
+    // Match up commitments with challenges
+    for challenge in challenges {
+        sectorsSorted[i] = sectors[challenge.Sector]
+    }
+
+    // Verify snark
+    VerifyPoStSnark(sectorSize, challenges, sectorsSorted)
+}
+```
+
+
+#### Types
+
+```go
+// The final proof stored on chain.
+type PoStProof struct {
+    snark []byte
+}
+```
+
+```go
+// The random challenge seed, provided by the chain.
+Seed [32]byte
+```
+
+```go
+type Challenge struct {
+    Sector Uint
+    Leaf Uint
+}
+```
+
+#### Challenge Derivation
+
+```go
+// Derive the full set of challenges for PoSt.
+func DerivePoStChallenges(seed Seed, faults FaultSet, sectorSize: Uint, sectorCount: Uint) [POST_CHALLENGES_COUNT]Challenge {
+    challenges := []
+
+    for n in 0..POST_CHALLENGES_COUNT {
+        attempt := 0
+        while challenges[n] == nil {
+            challenge := DerivePoStChallenge(seed, n, faults, attempt, sectorSize, sectorCount)
+
+            // check if we landed in a faulty sector
+            if !faults.Contains(challenge.Sector) {
+                // Valid challenge
+                challenges[n] = challenge
+            }
+            // invalid challenge, regenerate
+            attempt += 1
+        }
+    }
+
+    return challenges
+}
+
+// Derive a single challenge for PoSt.
+func DerivePoStChallenge(seed Seed, n Uint, attempt Uint, sectorSize Uint, sectorCount: Uint) Challenge {
+    n_bytes := WriteUintToLittleEndian(n)
+    data := concat(seed, n_bytes, WriteUintToLittleEndian(attempt))
+    challenge_bytes := blake2b(data)
+
+    sector_challenge := ReadUintLittleEndian(challenge_bytes[0..8])
+    leaf_challenge := ReadUintLittleEndian(challenge_bytes[8..16])
+
+    Challenge {
+        Sector: sector_challenge % sectorCount,
+        Leaf: leaf_challenge % (sectorSize / NODE_SIZE),
+    }
+}
+```
+
+
+### PoSt Circuit
+
+#### Public Parameters
+
+*Parameters that are embeded in the circuits or used to generate the circuit*
+
+- `POST_CHALLENGES_COUNT: UInt`: Number of challenges.
+- `POST_TREE_DEPTH: UInt`: Depth of the Merkle tree. Note, this is `(log_2(Size of original data in bytes/32 bytes per leaf))`.
+- `SECTOR_SIZE: UInt`:
+
+#### Public Inputs
+
+*Inputs that the prover uses to generate a SNARK proof and that the verifier uses to verify it*
+
+- `CommRs: [POST_CHALLENGES_COUNT]Fr`: The Merkle tree root hashes of all CommRs, ordered to match the inclusion paths and challenge order.
+- `InclusionPaths: [POST_CHALLENGES_COUNT]Fr`: Inclusion paths for the replica leafs, ordered to match the `CommRs` and challenge order. (Binary packed bools)
```

Comment:
> (Binary packed bools)

Is this detail relevant to this part of the spec? If not, I suggest terminating it. If so, ignore me.

Reply:
It is relevant and explained in more detail in the zigzag spec I believe